### ISSUE TITLE
Pushing error branch tip

### ISSUE THREAD

**Issue by slabzzz** [denolehov/obsidian-git#224]
When pulling I get a message like this: Failed to push some refs to ... updates were rejected cuz the tip of branch is behind its remote counterpart. Then it asks me to git pull which I did and it said everything is up to date. But the error persists.

Corresponding messages:

<img width="288" alt="Screen Shot 2022-04-30 at 10 01 36 PM" src="https://user-images.githubusercontent.com/100169858/166132923-70bdbdf2-2fee-48a3-9e11-3c09eb03c8d4.png">
.
<img width="195" alt="image" src="https://user-images.githubusercontent.com/100169858/166132933-b8b414e7-887a-4238-bdcc-dca184cc3214.png">

**Comment by slabzzz** [comment]
I accidentally edited remotes and now it is giving me these errors

https://user-images.githubusercontent.com/100169858/166160803-f2e6c793-7105-4be7-a4c0-c1b8a2271b22.mp4

:

**Comment by Vinzent03** [comment]
What is your sync method? E.g. `merge` or `rebase`

**Comment by slabzzz** [comment]
It is merge. Maybe if I rebase it it will be better? My only goal is to upload to github as backups. (I have no idea how to actually use git).

**Comment by ArchieAtkinson** [comment]
I am also getting this error, pushing with the command line works fine.

**Comment by Vinzent03** [comment]
I'm closing this for now. If you still have the same problem and use the newest plugin version, I will reopen it.

**Comment by itszn** [comment]
I repro'ed and tracked this issue down.
It happens when you have a repository that is behind its origin but does not have a remote tracking branch set.
This is because `src/simpleGit.ts` uses `git push --set-upsteam <remote>` to set the tracking branch. This will fail when upstream is ahead of the current branch:
https://github.com/denolehov/obsidian-git/blob/ffc76c281cf67b4a85bc21d377d79677004025b4/src/simpleGit.ts#L383

Here is my fix @Vinzent03:
Use `git branch --set-upstream-to <remote> <branch>` instead of using `git push --set-upstream <remote>`

Note, that flag used to be `branch --set-upsteam` until git 1.8.0.

Here is a patch that falls back to the old behavior on older versions:
```patch
diff --git a/src/simpleGit.ts b/src/simpleGit.ts
index ffc76c2..129b596 100644
--- a/src/simpleGit.ts
+++ b/src/simpleGit.ts
@@ -380,8 +380,19 @@ export class SimpleGit extends GitManager {
     }

     async updateUpstreamBranch(remoteBranch: string) {
-        await this.git.push(["--set-upstream", ...remoteBranch.split("/")], (err) => this.onError(err));
-
+        const branchInfo = await this.branchInfo();
+        try {
+            // git 1.8+
+            await this.git.branch(['--set-upstream-to', remoteBranch, branchInfo.current], (err) => this.onError(err));
+        } catch (e) {
+            try {
+                // git 1.7 - 1.8
+                await this.git.branch(['--set-upstream', remoteBranch, branchInfo.current], (err) => this.onError(err));
+            } catch(e) {
+                // fallback
+                await this.git.push(["--set-upstream", ...remoteBranch.split("/")], (err) => this.onError(err));
+            }
+        }
     }

     updateGitPath(gitPath: string) {
`

**Comment by Vinzent03** [comment]
@itszn Oh you are right, this may be the issue. Thanks for the code snippet. Available in `1.31.0`
